File: app/src/main/java/com/alvarlagerlof/koda/Api/ApiGetData.java
package com.alvarlagerlof.koda.Api;

import android.content.Context;
import android.os.AsyncTask;
import android.support.v7.widget.RecyclerView;

import com.alvarlagerlof.koda.ConnectionUtils;
import com.alvarlagerlof.koda.Cookies.PersistentCookieStore;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.net.CookieHandler;
import java.net.CookieManager;
import java.net.CookiePolicy;
import java.util.ArrayList;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.internal.JavaNetCookieJar;

/**
 * Created by alvar on 2016-11-08.
 */

class ApiGetData extends AsyncTask<Void, Void, String> {

    private Context context;
    private String url;
    private ArrayList<ApiObject> list;
    private RecyclerView.Adapter adapter;

    ApiGetData(Context context, String url, ArrayList<ApiObject> list, RecyclerView.Adapter adapter) {
        this.context = context;
        this.url = url;
        this.list = list;
        this.adapter = adapter;
    }




    // Code here
    @Override
    final protected void onPreExecute() {

        list.add(new ApiObject("Loading", ""));
        adapter.notifyDataSetChanged();

    }

    @Override
    final protected String doInBackground(Void... progress) {
        if (ConnectionUtils.isConnected(context)) {
            try {

                CookieHandler cookieHandler = new CookieManager(
                        new PersistentCookieStore(context), CookiePolicy.ACCEPT_ALL);

                OkHttpClient client = new OkHttpClient.Builder()
                        .cookieJar(new JavaNetCookieJar(cookieHandler))
                        .build();

                Request request = new Request.Builder()
                        .url(url)
                        .build();

                Response response = client.newCall(request).execute();
                String json = response.body().string();
                response.body().close();

                return json;

            } catch (Exception e) {
                e.printStackTrace();
            }

        }

        return null;
    }


    @Override
    final protected void onPostExecute(String json) {

        if (json != null) {

            try {
                list.remove(0);

                JSONArray jsonArray = new JSONArray(json);

                for (int i = 0; i < jsonArray.length(); i++) {
                    try {
                        JSONObject jsonobject = jsonArray.getJSONObject(i);
                        String command = jsonobject.getString("command");
                        String description = jsonobject.getString("description");
                        list.add(new ApiObject(command, description));
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }

                adapter.notifyDataSetChanged();

            } catch (JSONException e) {
                e.printStackTrace();
            }

        }
    }
}

File: app/src/main/java/com/alvarlagerlof/koda/Api/ApiGetData.kt
package com.alvarlagerlof.koda.Api

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import com.alvarlagerlof.koda.Cookies.PersistentCookieStore
import com.alvarlagerlof.koda.Extensions.isConnected
import com.alvarlagerlof.koda.Universal.UniversalLoadingObject
import com.alvarlagerlof.koda.Universal.UniversalOfflineObject
import com.alvarlagerlof.koda.Vars
import com.arasthel.asyncjob.AsyncJob
import com.google.firebase.analytics.FirebaseAnalytics
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.internal.JavaNetCookieJar
import org.json.JSONArray
import org.json.JSONException
import java.net.CookieManager
import java.net.CookiePolicy
import java.util.*

/**
 * Created by alvar on 2016-11-08.
 */

class ApiGetData(private val context: Context, private val url: String, private val list: ArrayList<Any>, private val adapter: RecyclerView.Adapter<*>) {

    init {
        list.add(UniversalLoadingObject())
        adapter.notifyDataSetChanged()

        AsyncJob.doInBackground {
            if (context.isConnected()) {
                try {

                    val cookieHandler = CookieManager(PersistentCookieStore(context), CookiePolicy.ACCEPT_ALL)
                    val client = OkHttpClient.Builder().cookieJar(JavaNetCookieJar(cookieHandler)).build()

                    val request = Request.Builder().url(url).build()

                    val response = client.newCall(request).execute()
                    val json = response.body()!!.string()
                    response.body()!!.close()

                    if (json != null) {
                        try {
                            list.clear()

                            val jsonArray = JSONArray(json)

                            for (i in 0..jsonArray.length() - 1) {
                                try {
                                    val jsonobject = jsonArray.getJSONObject(i)
                                    val command = jsonobject.getString("command")
                                    val description = jsonobject.getString("description")

                                    list.add(ApiObject(command, description))
                                } catch (e: JSONException) {
                                    e.printStackTrace()
                                }

                            }

                            AsyncJob.doOnMainThread { adapter.notifyDataSetChanged() }

                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }
                    } else {
                        list.clear()
                        list.add(UniversalOfflineObject())
                        AsyncJob.doOnMainThread { adapter.notifyDataSetChanged() }
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                }

            } else {
                list.clear()
                list.add(UniversalOfflineObject())
                AsyncJob.doOnMainThread { adapter.notifyDataSetChanged() }

                when (url) {
                    Vars.URL_API_2D -> FirebaseAnalytics.getInstance(context).logEvent("api_tab_2d", Bundle())
                    Vars.URL_API_3D -> FirebaseAnalytics.getInstance(context).logEvent("api_tab_3d", Bundle())
                }

            }
        }
    }
}